presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcess.java

```diff
@@ -0,0 +1,265 @@
+    private final HttpClient httpClient;
+    private final NativeExecutionTask nativeExecutionTask;
+    private Process process;
+
+    public NativeExecutionProcess(
+            Session session,
+            URI uri,
+            TaskId taskId,
+            PlanFragment planFragment,
+            List<TaskSource> sources,
+            TableWriteInfo tableWriteInfo,
+            HttpClient httpClient,
+            Executor executor,
+            ScheduledExecutorService errorRetryScheduledExecutor,
+            ScheduledExecutorService updateScheduledExecutor,
+            JsonCodec<TaskInfo> taskInfoCodec,
+            JsonCodec<PlanFragment> planFragmentCodec,
+            JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec,
+            JsonCodec<ServerInfo> serverInfoCodec,
+            Duration maxErrorDuration,
+            TaskManagerConfig taskManagerConfig,
+            NativeExecutionSystemConfig systemConfig,
+            NativeExecutionNodeConfig nodeConfig,
+            NativeExecutionConnectorConfig connectorConfig)
+            throws IOException
+    {
+        this.port = getAvailableTcpPort();
+        this.session = requireNonNull(session, "session is null");
+        this.location = getBaseUriWithPort(requireNonNull(uri, "uri is null"), port);
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+        this.workerClient = new PrestoSparkHttpWorkerClient(
+                this.httpClient,
+                taskId,
+                location,
+                taskInfoCodec,
+                planFragmentCodec,
+                taskUpdateRequestCodec,
+                serverInfoCodec,
+                taskManagerConfig.getInfoRefreshMaxWait());
+        this.taskManagerConfig = requireNonNull(taskManagerConfig, "taskManagerConfig is null");
+        this.systemConfig = requireNonNull(systemConfig, "systemConfig is null");
+        this.nodeConfig = requireNonNull(nodeConfig, "nodeConfig is null");
+        this.connectorConfig = requireNonNull(connectorConfig, "connectorConfig is null");
+        this.nativeExecutionTask = new NativeExecutionTask(session, planFragment, sources, this.workerClient, tableWriteInfo, executor, updateScheduledExecutor, taskManagerConfig);
+        this.errorRetryScheduledExecutor = requireNonNull(errorRetryScheduledExecutor, "errorRetryScheduledExecutor is null");
+        this.errorTracker = nativeExecutionRequestErrorTracker("NativeExecution", uri, maxErrorDuration, errorRetryScheduledExecutor, "getting native process status");
+    }
+
+    /**
+     * Starts the external native execution process. The method will be blocked by connecting to the native process's /v1/info endpoint with backoff retries until timeout.
+     */
+    public void start()
+            throws ExecutionException, InterruptedException, IOException
+    {
+        String executablePath = getProcessWorkingPath(SystemSessionProperties.getNativeExecutionExecutablePath(session));
```

Comment:
Ensure that this property can never be overriden and only be set by the system admins. Otherwise, bad actors might trick Presto in to running untrusted binaries. 

It will also be good, if you ensure that the file exists with this executable and fail with a meaningful error when the file is not present.

Reply:
That's a good point. On spark side (internally) we prevent the bad actor that by not distributing untrusted libraries, so even user set this property, the library won't be distributed to the worker for execution.

We also already have the file not exist check inside the `getProcessWorkingPath` method.